### PR TITLE
[JENKINS-64510] Issues with TCP agent reconnects

### PR DESCRIPTION
**Pull request by jeffret-b**
See [JENKINS-64510](https://issues.jenkins.io/browse/JENKINS-64510)

The reconnect mechanism for TCP inbound agents hasn't been working properly for some period of time, most likely a long time. It's not a serious loss, but getting it working again can make some minor improvements in clarity and robustness.

This is a necessary, but insufficient, change to get it working. A corresponding change is also necessary in Jenkins core. Both are required to get it working correctly, but they're loosely coupled otherwise. Individually, they don't make the situation any worse, so it's fine to only have one of them.

### CODE REVIEW

File: src/main/java/org/jenkinsci/remoting/engine/JnlpConnectionState.java [hunk 1]
```diff
@@ -66,7 +66,7 @@
     /**
      * The proprty name for the cookie name key.
      */
-    public static final String COOKIE_KEY = "Cookie";
+    public static final String COOKIE_KEY = "JnlpAgentProtocol.cookie";
```
Comment (jeffret-b) [inline src/main/java/org/jenkinsci/remoting/engine/JnlpConnectionState.java:69]: Use a more interesting, common name between agent and controller.

File: src/main/java/org/jenkinsci/remoting/engine/JnlpProtocol4Handler.java [hunk 2]
```diff
@@ -152,7 +152,7 @@ public Jnlp4ConnectionState createConnectionState(@Nonnull Socket socket,
                 .filter(new ConnectionHeadersFilterLayer(headers, handler))
                 .named(String.format("%s connection from %s", getName(), socket.getRemoteSocketAddress()))
                 .listener(handler)
-                .build(new ChannelApplicationLayer(threadPool, handler))
+                .build(new ChannelApplicationLayer(threadPool, handler, headers.get(JnlpConnectionState.COOKIE_KEY)))
```
Comment (jeffret-b) [inline src/main/java/org/jenkinsci/remoting/engine/JnlpProtocol4Handler.java:155]: This part and the corresponding additions to ChannelApplicationLayer are part of attaching the cookie to the channel properties on the controller side.

**Comment by basil** [comment]
`hudson.plugins.swarm.PipelineJobRestartTest` verifies that a Swarm agent can successfully reconnect to a Jenkins controller after a restart. Would be good to confirm that this functionality in Swarm remains unaffected by this proposed change.

**Comment by jeffret-b** [comment]
Yes, I certainly don't want to break that. Is that something I can easily run in a command-line or CI environment? Or otherwise, could you help test it when I have everything prepared.

**Comment by basil** [comment]
It's a standard integration test which you can run with `mvn verify` or in an IDE.

**Comment by jeffret-b** [comment]
Hmm ... I tried updating the Jenkins version in swarm-plugin to something recent (for example, 2.363) and it gets a compilation error in `GlobalSecurityConfigurationBuilder`.

**Comment by basil** [comment]
> I tried updating the Jenkins version in swarm-plugin to something recent (for example, 2.363) and it gets a compilation error in `GlobalSecurityConfigurationBuilder`

Presumably because jenkinsci/jenkins#4027 removed a public API from Jenkins core without updating pre-existing consumers.

CC @Wadeck

**Comment by jeffret-b** [comment]
See jenkinsci/jenkins#5138 for the corresponding core changes needed to tie these together.

**Comment by jeffret-b** [comment]
Maybe a rebuild will help.

**Comment by jeffret-b** [comment]
Incrementals worked this time. Currently at 4.7-rc2860.f4a9a135655d

**Comment by jeffret-b** [comment]
I ran the swarm build and all the tests passed based on the updated core PR so this should be safe there.

**Review by jvz — APPROVED** [review]

**Comment by jeffret-b** [comment]
I'm marking this as ready-for-merge. Any further additional reviews are welcome. I'll try to merge it in a day or two.

**Pull request merged**